src/utilities/schemaComparisons.js

```diff
@@ -0,0 +1,239 @@
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+import {
+  getNamedType,
+  GraphQLEnumType,
+  GraphQLInputObjectType,
+  GraphQLInterfaceType,
+  GraphQLObjectType,
+  GraphQLUnionType,
+} from '../type/definition';
+import {
+  GraphQLSchema,
+} from '../type/schema';
+
+export const BreakingChangeType = {
+  FIELD_CHANGED_KIND: 'FIELD_CHANGED_KIND',
+  FIELD_REMOVED: 'FIELD_REMOVED',
+  TYPE_CHANGED_KIND: 'TYPE_CHANGED_KIND',
+  TYPE_REMOVED: 'TYPE_REMOVED',
+  TYPE_REMOVED_FROM_UNION: 'TYPE_REMOVED_FROM_UNION',
+  VALUE_REMOVED_FROM_ENUM: 'VALUE_REMOVED_FROM_ENUM',
+};
+
+export type BreakingChange = {
+  type: $Keys<typeof BreakingChangeType>;
+  description: string;
+};
+
+/**
+ * Given two schemas, returns an Array containing descriptions of all the types
+ * of breaking changes covered by the other functions down below.
+ */
+export function findBreakingChanges(
+  oldSchema: GraphQLSchema,
+  newSchema: GraphQLSchema
+): Array<BreakingChange> {
+  return [
+    ...findRemovedTypes(oldSchema, newSchema),
+    ...findTypesThatChangedKind(oldSchema, newSchema),
+    ...findFieldsThatChangedType(oldSchema, newSchema),
+    ...findTypesRemovedFromUnions(oldSchema, newSchema),
+    ...findValuesRemovedFromEnums(oldSchema, newSchema)
+  ];
+}
+
+/**
+ * Given two schemas, returns an Array containing descriptions of any breaking
+ * changes in the newSchema related to removing an entire type.
+ */
+export function findRemovedTypes(
+  oldSchema: GraphQLSchema,
+  newSchema: GraphQLSchema
+): Array<BreakingChange> {
+  const oldTypes = Object.keys(oldSchema.getTypeMap());
+  const newTypes = new Set(Object.keys(newSchema.getTypeMap()));
+  return oldTypes.filter(typeName => !newTypes.has(typeName)).map(
+    type => {
+      return {
+        type: BreakingChangeType.TYPE_REMOVED,
+        description: `${type} was removed`,
+      };
+    }
+  );
+}
+
+/**
+ * Given two schemas, returns an Array containing descriptions of any breaking
+ * changes in the newSchema related to changing the type of a type.
+ */
+export function findTypesThatChangedKind(
+ oldSchema: GraphQLSchema,
+ newSchema: GraphQLSchema
+): Array<BreakingChange> {
+  const oldTypeMap = oldSchema.getTypeMap();
+  const newTypeMap = newSchema.getTypeMap();
+
+  const typesThatChangedType = [];
+  Object.keys(oldTypeMap).forEach(typeName => {
+    if (!newTypeMap[typeName]) {
+      return;
+    }
+    const oldType = oldTypeMap[typeName];
+    const newType = newTypeMap[typeName];
+    if (!(oldType instanceof newType.constructor)) {
+      typesThatChangedType.push(
+        {
+          type: BreakingChangeType.TYPE_CHANGED_KIND,
+          description: `${typeName} changed from a ` +
+            `${oldType.constructor.name} to a ${newType.constructor.name}`,
```

Comment:
These error names might be less clear to non JS users if this utility is used along with other servers.

This might print:

`Foo changed from a GraphQLObjectType to a GraphQLScalarType`

but might be easier to understand as:

`Foo changed from an Object type to a Scalar type`